src/commands/project.lisp

```diff
@@ -0,0 +1,231 @@
+                       ;; (try to) handle non-unix platforms path separators.
+                       (add-directory-separator marker) pathname)
+        if (member subpath subdirectories :test #'equal)
+          return t
+        finally (return)))
+
+(defvar *recurse* t)
+
+(defun find-root (pathname &key (recurse *recurse*) (recursing nil))
+  "Search for a root directory up the file hierarchy.
+  Return a pathname, or NIL.
+
+  root: a buffer directory (pathname).
+  recurse: if non T, do not recurse in parent directories (testing only)."
+  ;; recursing: when calling find-root on the HOME directory, it's OK to return it as the root.
+  ;; Otherwise, return nil (to get the caller directory).
+  (check-type pathname (or pathname string))
+  (setf pathname (pathname pathname))
+  (cond
+    ;; Stop at /
+    ((or (equal (directory-namestring pathname)
+                "/")
+         ;; XXX: windows??
+         (equal (directory-namestring pathname)
+                (string (uiop:directory-separator-for-host))))
+     (if recursing nil pathname))
+    ;; Stop at HOME.
+    ((equal (directory-namestring pathname)
+            (directory-namestring (user-homedir-pathname)))
+     (if recursing nil pathname))
+    ;; Stop if found a directory root.
+    ((found-subdirectory-root-p pathname)
+     pathname)
+    ;; Stop if found a file root.
+    ((found-file-root-p pathname)
+     pathname)
+    ;; Go up to the parent directory.
+    (t
+     (when recurse
+       (or
+        (find-root (parent-directory pathname) :recurse recurse :recursing t)
+        pathname)))))
+
+(define-command project-find-file (arg) ("p")
+  "Open a file, from the list of all files in this project."
+  ;; ARG is currently not used, use it when needed.
+  (declare (ignorable arg))
+  (let* ((cwd (buffer-directory))
+         (project-root (find-root cwd))
+         (root (or project-root cwd)))
+    (uiop:with-current-directory (root)
+      (let ((filename (prompt-for-files-recursively))
+            buffer)
+        (when filename
+          (setf buffer (execute-find-file *find-file-executor*
+                                          (get-file-mode filename)
+                                          filename))
+          (when buffer
+            (switch-to-buffer buffer t nil)))))))
+
+(define-command project-root () ()
+  "Display this buffer's project directory."
+  (let* ((cwd (buffer-directory))
+         (project-root (find-root cwd))
+         (root (or project-root cwd)))
+    (message "Current project root: ~a" root)))
+
+(define-command project-root-directory (arg) ("p")
+  "Open this project's root directory."
+  (declare (ignorable arg))
+  (let* ((cwd (buffer-directory))
+         (project-root (find-root cwd))
+         (root (or project-root cwd)))
+    (find-file root)))
+
+
+
+(defun %buffer-list ()
+  ;; Defined after this file in the .asd, need to defer the call.
+  (uiop:symbol-call :lem/list-buffers 'buffer-list))
```

Comment:
I think this is fine simply with `buffer-list`.
This is because buffer-list exists in lem-base.

Reply:
Oh you are right again. I forgot what lead me to this.